tests/integration/encryption/peer_share_test.go

```diff
@@ -0,0 +1,407 @@
+				Schema: `
+					type User {
+						name: String
+						age: Int
+					}
+				`,
+			},
+			testUtils.ConnectPeers{
+				SourceNodeID: 1,
+				TargetNodeID: 0,
+			},
+			testUtils.SubscribeToCollection{
+				NodeID:        1,
+				CollectionIDs: []int{0},
+			},
+			testUtils.CreateDoc{
+				NodeID:          immutable.Some(0),
+				Doc:             john21Doc,
+				IsDocEncrypted:  true,
+				EncryptedFields: []string{"name", "age"},
+			},
+			testUtils.WaitForSync{
+				Event:   immutable.Some(encryption.KeysRetrievedEventName),
+				NodeIDs: []int{1},
+			},
+			testUtils.Request{
+				NodeID: immutable.Some(1),
+				Request: `query {
+					User {
+						name
+						age
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"name": "John",
+							"age":  int64(21),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestDocEncryptionPeer_IfAllFieldsOfPublicDocAreIndividuallyEncrypted_ShouldFetchKeysAndDecrypt(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			testUtils.RandomNetworkingConfig(),
+			testUtils.RandomNetworkingConfig(),
+			testUtils.SchemaUpdate{
+				Schema: `
+					type User {
+						name: String
+						age: Int
+					}
+				`,
+			},
+			testUtils.ConnectPeers{
+				SourceNodeID: 1,
+				TargetNodeID: 0,
+			},
+			testUtils.SubscribeToCollection{
+				NodeID:        1,
+				CollectionIDs: []int{0},
+			},
+			testUtils.CreateDoc{
+				NodeID:          immutable.Some(0),
+				Doc:             john21Doc,
+				EncryptedFields: []string{"name", "age"},
+			},
+			testUtils.WaitForSync{
+				Event:   immutable.Some(encryption.KeysRetrievedEventName),
+				NodeIDs: []int{1},
+			},
+			testUtils.Request{
+				NodeID: immutable.Some(1),
+				Request: `query {
+					User {
+						name
+						age
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"name": "John",
+							"age":  int64(21),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestDocEncryptionPeer_WithUpdatesOnEncryptedDeltaBasedCRDTField_ShouldDecryptAndCorrectlyMerge(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			testUtils.RandomNetworkingConfig(),
+			testUtils.RandomNetworkingConfig(),
+			testUtils.SchemaUpdate{
+				Schema: `
+					type User {
+						name: String
+						age: Int @crdt(type: "pcounter")
+					}
+				`,
+			},
+			testUtils.ConnectPeers{
+				SourceNodeID: 1,
+				TargetNodeID: 0,
+			},
+			testUtils.CreateDoc{
+				NodeID:          immutable.Some(0),
+				Doc:             john21Doc,
+				EncryptedFields: []string{"age"},
+			},
+			testUtils.UpdateDoc{
+				NodeID: immutable.Some(0),
+				Doc:    `{"age": 3}`,
+			},
+			testUtils.SubscribeToCollection{
+				NodeID:        1,
+				CollectionIDs: []int{0},
+			},
+			testUtils.UpdateDoc{
+				NodeID: immutable.Some(0),
+				Doc:    `{"age": 2}`,
+			},
+			testUtils.WaitForSync{
```

Comment:
what do you mean "why"? The reason is the same as with all others peers-related tests.
No, if I remove it, it will not fail. But if you run other integration tests and comment out `WaitForSync` it will also pass.
It is there to make sure the test is not flaky.
Is your question specifically about `encryption.KeysRetrievedEventName` or sync in general?
And I'm not actually sure which new information documentation can add. It is pretty explicit: wait on node 1 for keys-retrieved event.

Reply:
> It is pretty explicit: wait on node 1 for keys-retrieved event.

Why does it need to wait for the keys-retrieved event?  What happens if the request is processed before the event?

Reply:
we wait for not 1 to received the encryption key. If we do it, node 1 won't be able to fetch decrypted doc in the request (following action)

I don't see why it needs to be commented.

Reply:
> If we don't do it, node 1 won't be able to fetch decrypted doc in the request (following action)

What happens if it tries to do so?

> I don't see why it needs to be commented.

It is far from obvious what the public behaviour is (I still dont know), and it could be a fairly common thing for users to come across.

Reply:
> What happens if it tries to do so? Does it error, panic, return an empty set, wait for the key-sync during the request?

If node 1 doesn't receive the enc key and makes gql request, the result will be empty. I actually had a test that shows this in my previous PR that I remove in this PR 
https://github.com/sourcenetwork/defradb/pull/2891/files#diff-6fd6bae6910d2da0509cd14c385e002a3c3969665086213ed36bc37d28b2927cL21-L56

In the previous PR there was no key exchange, so all keys were treated as private, but the DAG anyway was synced.
With this PR all docs are assumed to be public because a node gives keys to whoever requests.

Once we are done with permission granting they can be married together.

Reply:
`AnyOf` is supported only for doc's field values, not for the entire doc. 

But the idea with test is nice. Anyway we will have bunch of tests like this once permission sharing is in place

Reply:
brought back the test with your `AnyOf` suggestion. Had to tweak our testing framework a bit